www/source/blog/2017-10-24-node-with-db.html.md

```diff
@@ -0,0 +1,348 @@
+
+**habitat/default.toml**
+
+```
+# Use this file to templatize your application's native configuration files.
+# See the docs at https://www.habitat.sh/docs/create-packages-configure/.
+# You can safely delete this file if you don't need it.
+
+[mongo]
+host = "127.0.0.1"
+port = "27017"
+```
+
+Save and close that file.
+
+Now let's open up the plan file that was created when we ran "hab init", it should look like this:
+
+**habitat/plan.sh**
+
+```
+pkg_name=learn-about-me
+pkg_origin=core
+pkg_version="0.1.0"
+pkg_scaffolding="core/scaffolding-node"
+```
+
+When we eventually start this application in a container, it will need to bind to the other container running MongoDB.  We need a port to bind them.  Add this content to your plan.sh.
+
+**habitat/plan.sh**
+
+```
+pkg_name=learn-about-me
+pkg_origin=core
+pkg_version="0.1.0"
+pkg_scaffolding="core/scaffolding-node"
+
+pkg_binds=(
+  [database]="port"
+)
+```
+
+Now, let's define the value for that port in a configuration file - this will tell Habitat where to connect to the database when binding the two containers (unlike habitat/default.toml - which is used to set environmental variables within the application itself).
+
+**default-config.json**
+
+```
+{
+  "mongo": {
+    "host" : "localhost",
+    "port"   : "27017"
+  }
+}
+```
+
+Let's add one more thing to the plan.sh file - when we run the application in a Docker container, we need a port to access it on.
+
+**habitat/plan.sh**
+
+```
+pkg_name=learn-about-me
+pkg_origin=core
+pkg_version="0.1.0"
+pkg_scaffolding="core/scaffolding-node"
+
+pkg_binds=(
+  [database]="port"
+)
+
+pkg_exports=(
+  [port]=app.port
+)
+```
+
+We will define this port when we start up the container running our application - we'll get to that in just a moment.
+
+## Habitizing your application - Part II
+
+We need to make a few edits to the application code itself in order to make this work with Habitat. Open up your app.js file, the beginning of it should look like this:
+
+**app.js**
+
+```
+var bodyParser = require("body-parser");
+var cookieParser = require("cookie-parser");
+var express = require("express");
+var flash = require("connect-flash");
+var mongoose = require("mongoose");
+var passport = require("passport");
+var path = require("path");
+var session = require("express-session");
+
+var setUpPassport = require("./setuppassport");
+var routes = require("./routes");
+
+var app = express();
+mongoose.connect("mongodb://localhost:27017/test");
+setUpPassport();
+```
+
+First, delete this line - since we define the database url through Habitat:
+
+```
+mongoose.connect("mongodb://localhost:27017/test");
+```
+
+Now, add some content right after this line:
+
+```
+var app = express();
+```
+
+Like so:
+
+**app.js**
+
+```
+var app = express();
+var nconf = require('nconf');
+const nconf_file = process.env.APP_CONFIG || './default-config.json';
```

Comment:
The first part of the this statement is reading the environment variable which contains a file. That's the connection point into the application.

So you could create a file first and then ask the question - how do I set it as an environment variable during run time?
Or you could set the environment variable in the plan and then create the file.